File: ios/scanmynet_sdk/Sources/scanmynet_sdk/ReportMapper.swift
import Foundation
import ScanMyNet

/// Maps the native `ScanMyNet` SDK's `SMN*` report models onto the Pigeon report
/// types (`ReportData` and its sections).
///
/// This is the iOS counterpart of the Android `ReportMapper.kt`; both feed the
/// same Pigeon `ScanResult.report`, so the two must stay field-for-field aligned.
///
/// Two rules run throughout:
///
///   * **Null in, null out** — an absent section stays absent rather than becoming
///     an empty object, so Dart can tell "not measured" from "measured empty".
///   * **`Int` -> `Int64` widening** — Pigeon models Dart `int` as Swift `Int64`,
///     while the SDK models use the native `Int`. Every integer field is widened
///     with `{ Int64($0) }`; omitting this is a compile error, not a silent bug.
extension SMNReportPayload {
  func toPigeon() -> ReportData {
    ReportData(
      customerDetails: customerDetails?.toPigeon(),
      sdkDetails: sdkDetails?.toPigeon(),
      routerUsageDuringScan: routerUsageDuringScan?.toPigeon(),
      customerInternetSpeed: customerInternetSpeed?.toPigeon(),
      basicConnectivity: basicConnectivity?.toPigeon(),
      localConnectedDevices: localConnectedDevices?.map { $0.toPigeon() },
      customerRouterDetails: customerRouterDetails?.toPigeon(),
      networkTopology: networkTopology?.toPigeon(),
      userWifiNetwork: userWifiNetwork?.toPigeon(),
      networkCongestion: networkCongestion?.toPigeon(),
      connectionQuality: connectionQuality?.map { $0.toPigeon() },
      traceroute: traceroute?.map { $0.toPigeon() },
      alerts: alerts?.map { $0.toPigeon() },
      actions: actions?.map { $0.toPigeon() },
      incompleteAnalysis: incompleteAnalysis
    )
  }
}

extension SMNCustomerDetails {
  func toPigeon() -> CustomerDetails {
    CustomerDetails(
      key: key,
      lastKnownPublicIp: lastKnownPublicIP,
      // Capitalised keys (`Country`, `Region`, `ISP`) pass through verbatim.
      lastKnownPublicIpDetails: lastKnownPublicIPDetails
    )
  }
}

extension SMNSDKDetails {
  func toPigeon() -> SdkDetails {
    SdkDetails(
      start: start,
      duration: duration.map { Int64($0) },
      platform: platform,
      app: app,
      routeThisSdk: routeThisSDK,
      userPublicUpAddress: userPublicUpAddress,
      gpsLatitude: gpsLatitude,
      gpsLongitude: gpsLongitude
    )
  }
}

extension SMNRouterUsage {
  func toPigeon() -> RouterUsage {
    RouterUsage(networkUsageDown: networkUsageDown, networkUsageUp: networkUsageUp)
  }
}

extension SMNInternetSpeed {
  func toPigeon() -> InternetSpeed {
    InternetSpeed(
      networkSpeedDown: networkSpeedDown,
      networkSpeedUp: networkSpeedUp,
      currentNegotiatedLinkSpeed: currentNegotiatedLinkSpeed,
      maximumLinkSupportedByPhone: maximumLinkSupportedByPhone,
      networkSpeedUpSegments: networkSpeedUpSegments,
      networkSpeedDownSegments: networkSpeedDownSegments,
      // Single-entry map keyed by label — passed through as-is, NOT flattened.
      connectionQuality: connectionQuality
    )
  }
}

extension SMNBasicConnectivity {
  func toPigeon() -> BasicConnectivity {
    BasicConnectivity(
      ipAssignedViaDhcp: ipAssignedViaDHCP,
      serverConnectivity: serverConnectivity?.map { $0.toPigeon() },
      portChecks: portChecks?.map { $0.toPigeon() },
      dnsLookup: dnsLookup?.map { $0.toPigeon() },
      summary: summary?.toPigeon(),
      blockedServers: servers,
      blockedUdpPorts: udp?.map { Int64($0) },
      blockedTcpPorts: tcp?.map { Int64($0) },
      firewall: firewall?.toPigeon(),
      clientIsolation: clientIsolation?.toPigeon(),
      multicast: multicast?.toPigeon()
      // basic_connectivity.alerts is always empty — intentionally dropped, as on Android.
    )
  }
}

extension SMNServerConnectivityResult {
  func toPigeon() -> ServerConnectivityResult {
    ServerConnectivityResult(name: name, serverStatus: serverStatus)
  }
}

extension SMNPortCheckResult {
  func toPigeon() -> PortCheckResult {
    PortCheckResult(
      port: port.map { Int64($0) },
      portType: portType,
      description: description,
      portStatus: portStatus
    )
  }
}

extension SMNDNSLookupResult {
  func toPigeon() -> DnsLookupResult {
    DnsLookupResult(dnsIp: dnsIP, alias: alias, reverseDns: reverseDNS)
  }
}

extension SMNConnectivitySummary {
  func toPigeon() -> ConnectivitySummary {
    ConnectivitySummary(
      serverConnectivity: serverConnectivity?.toPigeon(),
      portChecks: portChecks?.toPigeon(),
      dnsLookup: dnsLookup
    )
  }
}

extension SMNStatusColor {
  func toPigeon() -> StatusColor { StatusColor(status: status, color: color) }
}

extension SMNToggleState {
  func toPigeon() -> ToggleState {
    // SMN `alerts` is a single object (not a list) — maps to the singular `alert`.
    ToggleState(value: value, color: color, status: status, alert: alerts?.toPigeon())
  }
}

extension SMNReportAlert {
  func toPigeon() -> ReportAlert {
    ReportAlert(alertType: alertType, alertValue: alertValue)
  }
}

extension SMNReportAction {
  func toPigeon() -> ReportAction {
    ReportAction(actionType: actionType, actionValue: actionValue)
  }
}

extension SMNLocalConnectedDevice {
  func toPigeon() -> LocalConnectedDevice {
    LocalConnectedDevice(
      deviceName: deviceName,
      deviceIp: deviceIP,
      deviceMacAddress: deviceMacAddress,
      manufacturer: manufacturer,
      packetsDropped: packetsDropped,
      numPacketsSent: numPacketsSent.map { Int64($0) },
      pingValues: pingValues,
      isSubscriberPhone: isSubscriberPhone,
      averagePingTime: averagePingTime,
      connectionQualityColor: connectionQualityColor?.toPigeon(),
      isSubscriberRouter: isSubscriberRouter,
      deviceDetails: deviceDetails?.toPigeon()
    )
  }
}

extension SMNQualityColor {
  func toPigeon() -> QualityColor { QualityColor(quality: quality, color: color) }
}

extension SMNDeviceRecognition {
  func toPigeon() -> DeviceRecognition {
    // `recognition` is an open payload: unwrap each AnyCodable to its raw value
    // so the map becomes the `[String: Any?]` the channel codec expects.
    DeviceRecognition(mac: mac, recognition: recognition?.mapValues { $0.value })
  }
}

extension SMNCustomerRouterDetails {
  func toPigeon() -> CustomerRouterDetails {
    CustomerRouterDetails(
      make: make,
      model: model,
      encryption: encryption,
      protocols: protocols,
      mesh: mesh,
      routerIpAddress: routerIPAddress,
      routerMacAddress: routerMacAddress,
      manufacturer: manufacturer,
      hostname: hostname,
      modelDescription: modelDescription,
      modelNumber: modelNumber,
      friendlyName: friendlyName,
      deviceType: deviceType,
      routerDetails: routerDetails?.toPigeon()
    )
  }
}

extension SMNNetworkTopology {
  func toPigeon() -> NetworkTopology {
    NetworkTopology(
      routerIpAddress: routerIPAddress,
      otherRouters: otherRouters,
      otherRoutersDetails: otherRoutersDetails?.map { $0.toPigeon() },
      doubleNatDetected: doubleNatDetected?.toPigeon()
    )
  }
}

extension SMNOtherRouterDetail {
  func toPigeon() -> OtherRouterDetail {
    OtherRouterDetail(ip: ip, asn: asn.map { Int64($0) }, owner: owner)
  }
}

extension SMNDoubleNat {
  func toPigeon() -> DoubleNat {
    DoubleNat(isDoubleNat: isDoubleNat, doubleNatHop: doubleNatHop)
  }
}

extension SMNWifiNetworkResult {
  func toPigeon() -> WifiNetworkResult {
    WifiNetworkResult(
      ssid: ssid,
      ssidIp: ssidIP,
      bssid: bssid,
      encryption: encryption,
      frequency: frequency,
      wpsAvailability: wpsAvailability,
      signalStrength: signalStrength.map { Int64($0) },
      numWifiChannels: numWifiChannels.map { Int64($0) },
      channelWidth: channelWidth.map { Int64($0) },
      currentChannel: currentChannel.map { Int64($0) },
      isSubscriberSsid: isSubscriberSSID
    )
  }
}

extension SMNNetworkCongestion {
  func toPigeon() -> NetworkCongestion {
    NetworkCongestion(
      userConnection: userConnection?.toPigeon(),
      environment: environment?.toPigeon()
    )
  }
}

extension SMNUserConnection {
  func toPigeon() -> UserConnection {
    UserConnection(
      phoneWifiFrequency: phoneWifiFrequency,
      numPhoneWifiChannel: numPhoneWifiChannel.map { Int64($0) },
      numNetworksOnChannel: numNetworksOnChannel.map { Int64($0) },
      phoneWifiChannel: phoneWifiChannel.map { Int64($0) }
    )
  }
}

extension SMNCongestionEnvironment {
  func toPigeon() -> CongestionEnvironment {
    CongestionEnvironment(
      channelCongestion: channelCongestion?.map { $0.toPigeon() },
      surroundingWifiNetworks: surroundingWifiNetworks?.map { $0.toPigeon() }
    )
  }
}

/// `index` is a String in the payload and stays one.
extension SMNChannelCongestion {
  func toPigeon() -> ChannelCongestion {
    ChannelCongestion(index: index, numNetworks: numNetworks.map { Int64($0) })
  }
}

extension SMNDNSQuality {
  func toPigeon() -> DnsQuality {
    DnsQuality(
      dnsName: dnsName,
      dnsIp: dnsIP,
      packetsDropped: packetsDropped,
      numPacketsSent: numPacketsSent.map { Int64($0) },
      pingValues: pingValues,
      isSubscriberRouter: isSubscriberRouter,
      jitter: jitter,
      averagePingTime: averagePingTime,
      connectionQualityColor: connectionQualityColor?.toPigeon(),
      layerRanking: layerRanking.map { Int64($0) }
    )
  }
}

extension SMNTracerouteResultEntry {
  func toPigeon() -> TracerouteEntry {
    TracerouteEntry(
      dnsDestinationIp: dnsDestinationIP,
      hops: hops?.map { $0.toPigeon() }
    )
  }
}

extension SMNTracerouteHop {
  func toPigeon() -> TracerouteHop {
    TracerouteHop(dnsIp: dnsIP, dnsName: dnsName, rttValues: rttValues?.map { Int64($0) })
  }
}

File: android/src/main/kotlin/com/creativeadvtech/scanmynet_sdk/ReportMapper.kt
package com.creativeadvtech.scanmynet_sdk

import com.creative.tools.rest.dto.response.*

/**
 * Maps the SDK's response DTOs onto the Pigeon report types.
 *
 * Null in, null out: an absent section stays absent rather than becoming an
 * empty object, so Dart can distinguish "not measured" from "measured empty".
 */
internal fun ReportDto.toPigeon() = ReportData(
    customerDetails = customer_details?.toPigeon(),
    sdkDetails = sdk_details?.toPigeon(),
    routerUsageDuringScan = router_usage_during_scan?.toPigeon(),
    customerInternetSpeed = customer_internet_speed?.toPigeon(),
    basicConnectivity = basic_connectivity?.toPigeon(),
    localConnectedDevices = local_connected_devices?.map { it.toPigeon() },
    customerRouterDetails = customer_router_details?.toPigeon(),
    networkTopology = network_topology?.toPigeon(),
    userWifiNetwork = user_wifi_network?.toPigeon(),
    networkCongestion = network_congestion?.toPigeon(),
    connectionQuality = connection_quality?.map { it.toPigeon() },
    traceroute = traceroute?.map { it.toPigeon() },
    alerts = alerts?.map { it.toPigeon() },
    actions = actions?.map { it.toPigeon() },
    incompleteAnalysis = incomplete_analysis,
)

internal fun CustomerDetailsDto.toPigeon() = CustomerDetails(
    key = key,
    lastKnownPublicIp = last_known_public_ip,
    // Capitalised keys pass through verbatim — no case transformation.
    lastKnownPublicIpDetails = last_known_public_ip_details,
)

internal fun SdkDetailsDto.toPigeon() = SdkDetails(
    start = start,
    duration = duration,
    platform = platform,
    app = app,
    routeThisSdk = route_this_sdk,
    userPublicUpAddress = user_public_up_address,
    gpsLatitude = gps_latitude,
    gpsLongitude = gps_longitude,
)

internal fun RouterUsageDto.toPigeon() = RouterUsage(
    networkUsageDown = network_usage_down,
    networkUsageUp = network_usage_up,
)

internal fun InternetSpeedDto.toPigeon() = InternetSpeed(
    networkSpeedDown = network_speed_down,
    networkSpeedUp = network_speed_up,
    currentNegotiatedLinkSpeed = current_negotiated_link_speed,
    maximumLinkSupportedByPhone = maximum_link_supported_by_phone,
    networkSpeedUpSegments = network_speed_up_segments,
    networkSpeedDownSegments = network_speed_down_segments,
    // Single-entry map keyed by label — passed through as-is, NOT flattened.
    connectionQuality = connection_quality,
)

internal fun BasicConnectivityDto.toPigeon() = BasicConnectivity(
    ipAssignedViaDhcp = ip_assigned_via_dhcp,
    serverConnectivity = server_connectivity?.map { it.toPigeon() },
    portChecks = port_checks?.map { it.toPigeon() },
    dnsLookup = dns_lookup?.map { it.toPigeon() },
    summary = summary?.toPigeon(),
    blockedServers = servers,
    blockedUdpPorts = udp,
    blockedTcpPorts = tcp,
    firewall = firewall?.toPigeon(),
    clientIsolation = client_isolation?.toPigeon(),
    multicast = multicast?.toPigeon(),
    // basic_connectivity.alerts is always empty — intentionally dropped.
)

internal fun ServerConnectivityResultDto.toPigeon() =
    ServerConnectivityResult(name = name, serverStatus = server_status)

internal fun PortCheckDto.toPigeon() = PortCheckResult(
    port = port, portType = port_type, description = description, portStatus = port_status,
)

internal fun DnsLookupResultDto.toPigeon() =
    DnsLookupResult(dnsIp = dns_ip, alias = alias, reverseDns = reverse_dns)

internal fun ConnectivitySummaryDto.toPigeon() = ConnectivitySummary(
    serverConnectivity = server_connectivity?.toPigeon(),
    portChecks = port_checks?.toPigeon(),
    dnsLookup = dns_lookup,
)

internal fun StatusColorDto.toPigeon() = StatusColor(status = status, color = color)

internal fun ToggleStateDto.toPigeon() = ToggleState(
    value = value, color = color, status = status, alert = alerts?.toPigeon(),
)

internal fun AlertDto.toPigeon() =
    ReportAlert(alertType = alert_type, alertValue = alert_value)

internal fun ActionDto.toPigeon() =
    ReportAction(actionType = action_type, actionValue = action_value)

internal fun LocalConnectedDeviceDto.toPigeon() = LocalConnectedDevice(
    deviceName = device_name,
    deviceIp = device_ip,
    deviceMacAddress = device_mac_address,
    manufacturer = manufacturer,
    packetsDropped = packets_dropped,
    numPacketsSent = num_packets_sent,
    pingValues = ping_values,
    isSubscriberPhone = is_subscriber_phone,
    averagePingTime = average_ping_time,
    connectionQualityColor = connection_quality_color?.toPigeon(),
    isSubscriberRouter = is_subscriber_router,
    deviceDetails = device_details?.toPigeon(),
)

internal fun QualityColorDto.toPigeon() = QualityColor(quality = quality, color = color)

internal fun DeviceRecognitionDto.toPigeon() =
    DeviceRecognition(mac = mac, recognition = recognition)

internal fun CustomerRouterDetailsDto.toPigeon() = CustomerRouterDetails(
    make = make, model = model, encryption = encryption, protocols = protocols,
    mesh = mesh, routerIpAddress = router_ip_address,
    routerMacAddress = router_mac_address, manufacturer = manufacturer,
    hostname = hostname, modelDescription = model_description,
    modelNumber = model_number, friendlyName = friendly_name,
    deviceType = device_type, routerDetails = router_details?.toPigeon(),
)

internal fun NetworkTopologyDto.toPigeon() = NetworkTopology(
    routerIpAddress = router_ip_address,
    otherRouters = other_routers,
    otherRoutersDetails = other_routers_details?.map { it.toPigeon() },
    doubleNatDetected = double_nat_detected?.toPigeon(),
)

internal fun OtherRouterDetailDto.toPigeon() =
    OtherRouterDetail(ip = ip, asn = asn, owner = owner)

internal fun DoubleNatDto.toPigeon() =
    DoubleNat(isDoubleNat = is_double_nat, doubleNatHop = double_nat_hop)

internal fun WifiNetworkResultDto.toPigeon() = WifiNetworkResult(
    ssid = ssid, ssidIp = ssid_ip, bssid = bssid, encryption = encryption,
    frequency = frequency, wpsAvailability = wps_availability,
    signalStrength = signal_strength, numWifiChannels = num_wifi_channels,
    channelWidth = channel_width, currentChannel = current_channel,
    isSubscriberSsid = is_subscriber_ssid,
)

internal fun NetworkCongestionDto.toPigeon() = NetworkCongestion(
    userConnection = user_connection?.toPigeon(),
    environment = environment?.toPigeon(),
)

internal fun UserConnectionDto.toPigeon() = UserConnection(
    phoneWifiFrequency = phone_wifi_frequency,
    numPhoneWifiChannel = num_phone_wifi_channel,
    numNetworksOnChannel = num_networks_on_channel,
    phoneWifiChannel = phone_wifi_channel,
)

internal fun CongestionEnvironmentDto.toPigeon() = CongestionEnvironment(
    channelCongestion = channel_congestion?.map { it.toPigeon() },
    surroundingWifiNetworks = surrounding_wifi_networks?.map { it.toPigeon() },
)

/** [ChannelCongestionDto.index] is a String and stays one. */
internal fun ChannelCongestionDto.toPigeon() =
    ChannelCongestion(index = index, numNetworks = num_networks)

internal fun DnsQualityDto.toPigeon() = DnsQuality(
    dnsName = dns_name, dnsIp = dns_ip, packetsDropped = packets_dropped,
    numPacketsSent = num_packets_sent, pingValues = ping_values,
    isSubscriberRouter = is_subscriber_router, jitter = jitter,
    averagePingTime = average_ping_time,
    connectionQualityColor = connection_quality_color?.toPigeon(),
    layerRanking = layer_ranking,
)

internal fun TracerouteResultDto.toPigeon() = TracerouteEntry(
    dnsDestinationIp = dns_destination_ip,
    hops = hops?.map { it.toPigeon() },
)

internal fun TracerouteHopDto.toPigeon() =
    TracerouteHop(dnsIp = dns_ip, dnsName = dns_name, rttValues = rtt_values)
